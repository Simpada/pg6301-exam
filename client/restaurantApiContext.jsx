import React from "react";
import {fetchJSON} from "./lib/fetchJSON";
import {postJSON} from "./lib/postJSON";


export const RestaurantApiContext = React.createContext({
    async fetchLogin() {
        return await fetchJSON("/api/login");
    },
    async menu(query) {
        return await fetchJSON("/api/menu?" + new URLSearchParams(query));
    },
    async addItem(item) {
        return await postJSON("/api/menu/add", item);
    }
})