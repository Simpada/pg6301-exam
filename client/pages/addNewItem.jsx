import {useContext, useState} from "react";
import {RestaurantApiContext} from "../restaurantApiContext";
import {useNavigate} from "react-router-dom";

export function AddNewItem() {

    const {addItem} = useContext(RestaurantApiContext);
    const [name, setName] = useState("");
    const [ingredients, setIngredients] = useState("");
    const [price, setPrice] = useState("");
    const [category, setCategory] = useState("");

    const navigate = useNavigate();

    function handleSubmit(e) {
        e.preventDefault();
        addItem({name, ingredients: ingredients.split(" "), price: parseInt(price), category});
        navigate("/");
    }

    return(
        <form onSubmit={handleSubmit}>
            <h1> Add new Item to The Menu </h1>

            <div className="form-input">
                <label>
                    <strong>Dish Name</strong>{" "}
                    <input value={name} onChange={(e) => setName(e.target.value)} />
                </label>
            </div>
            <div className="form-input">
                <label>

                    <strong>Ingredients</strong>{" "}
                    <input value={ingredients} onChange={(e) => setIngredients(e.target.value)} />
                    Separate the ingredients by spaces
                </label>
            </div>
            <div className="form-input">
                <label>
                    <strong>Price</strong>{" "}
                    <input
                        onKeyPress={(event) => {
                        if (!/[0-9]/.test(event.key)) {
                            event.preventDefault();
                        }}}
                        value={price} onChange={(e) => setPrice(e.target.value)} />
                </label>
            </div>
            <div className="form-input">
                <label>
                    <strong>Category</strong>{" "}
                    <input value={category} onChange={(e) => setCategory(e.target.value)} />
                </label>
            </div>
            <div>
                <button>Submit</button>
            </div>
        </form>
    )

}