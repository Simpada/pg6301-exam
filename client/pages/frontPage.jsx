import { Link } from "react-router-dom";
import React, { useContext } from "react";
import { useLoading } from "../useLoading";
import { RestaurantApiContext } from "../restaurantApiContext";

export function FrontPage(skip) {
  const { fetchLogin } = useContext(RestaurantApiContext);
  const { loading, error, data } = useLoading(async () => await fetchLogin());
  const user = data;

  let admin = false;

  if (skip.skip !== true) {
    if (loading) {
      return <div>Loading...</div>;
    }
    if (error) {
      return <div>{error.toString()}</div>;
    }
  }

  if (user !== null) {
    if (skip.skip !== true && user.isAdmin) {
      admin = true;
    }
  }

  return (
    <div>
      <h1>The API and Web-design Restaurant</h1>
      <ul>
        <li>
          <Link to={"/menu"}>View Menu</Link>
        </li>
        {admin ? (
          <li>
            <Link to={"/menu/add"}>Add an Item to The Menu</Link>
          </li>
        ) : (
          <div></div>
        )}
      </ul>
    </div>
  );
}
