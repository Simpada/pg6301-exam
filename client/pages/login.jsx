import { Link, useNavigate } from "react-router-dom";
import React, { useContext, useState } from "react";
import { RestaurantApiContext } from "../restaurantApiContext";

export function LoginLinks() {
  return (
    <div>
      <div>
        <Link to={"/login"}>Login</Link>
      </div>
      <div>
        <Link to={"/register"}>Register new user</Link>
      </div>
    </div>
  );
}

export function Login() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const navigate = useNavigate();

  const { login } = useContext(RestaurantApiContext);

  async function handleSubmit(e) {
    e.preventDefault();
    await login({ username, password });
    navigate("/");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Please log in</h1>
      <div>
        Username:{" "}
        <input value={username} onChange={(e) => setUsername(e.target.value)} />
      </div>
      <div>
        Password:{" "}
        <input
          type="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <div>
        <button>Log in</button>
      </div>
      <pre>{JSON.stringify({ username, password }, undefined, "  ")}</pre>
    </form>
  );
}
