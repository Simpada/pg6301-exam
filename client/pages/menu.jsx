import {useLoading} from "../useLoading";
import React, {useContext, useState} from "react";
import {RestaurantApiContext} from "../restaurantApiContext";


function ItemEntry({item: {name, ingredients, price, category}}) {
    return <><h3>{name}</h3>
        <div>Ingredients: {ingredients.map(e => `${e}, `)}</div>
        <div>Price: {price}</div>
        <div>Category: {category}</div>
    </>
}

export function Menu() {

    const {menu} = useContext(RestaurantApiContext)
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");
    const [price, setPrice] = useState("");
    const [priceQuery, setPriceQuery] = useState("");
    const {loading, error, data} = useLoading(
        async () => await menu({category, price}),
        [category, price]
    );


    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(categoryQuery);
        setPrice(priceQuery);
    }


    if (loading){
        return <div>Loading...</div>
    }
    if (error) {
        return <div>
            <h1>ERROR</h1>
            <div id="error-text">{error.toString()}</div>
        </div>
    }

    return <div>
        <h1> The Grand Menu </h1>

        <div>
            <form onSubmit={handleSubmitQuery}>
                <label>
                    Category:
                    <input
                        id="category-query"
                        value={categoryQuery}
                        onChange={(e) => setCategoryQuery(e.target.value)}
                    />
                </label>
                <br/>
                <label>
                    Price:
                    <input
                        id="price-query"
                        onKeyPress={(event) => {
                            if (!/[0-9]/.test(event.key)) {
                                event.preventDefault();
                            }
                        }}
                        value={priceQuery}
                        onChange={(e) => setPriceQuery(e.target.value)}
                    />
                </label>
                <br/>
                <button>Filter</button>
            </form>
        </div>

        {data.map((item) => (
            <ItemEntry key={item.name} item={item}/>
        ))}

    </div>;

}