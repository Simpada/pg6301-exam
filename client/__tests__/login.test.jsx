import * as ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { Login } from "../pages/login";
import React from "react";
import { RestaurantApiContext } from "../restaurantApiContext";
import { Simulate } from "react-dom/test-utils";

describe("login page component", () => {
  it("shows login form", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form strong")).map(
        (e) => e.innerHTML
      )
    ).toEqual(["Username:", "Password:"]);
  });

  it("logs in on submit", () => {
    const login = jest.fn();

    const element = document.createElement("div");

    ReactDOM.render(
      <RestaurantApiContext.Provider value={{ login }}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </RestaurantApiContext.Provider>,
      element
    );
    Simulate.change(element.querySelector(".form-input:nth-of-type(1) input"), {
      target: { value: "MrTest" },
    });
    Simulate.change(element.querySelector(".form-input:nth-of-type(2) input"), {
      target: { value: "TestPassword" },
    });
    Simulate.submit(element.querySelector("form"));

    expect(login).toBeCalledWith({
      username: "MrTest",
      password: "TestPassword",
    });
  });
});
