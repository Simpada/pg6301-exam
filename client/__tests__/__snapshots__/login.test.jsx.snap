// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login page component shows login form 1`] = `"<form><h1>Please log in</h1><div class="form-input"><strong>Username:</strong><input value=""></div><div class="form-input"><strong>Password:</strong><input type="password" value=""></div><div><button>Log in</button></div></form>"`;
